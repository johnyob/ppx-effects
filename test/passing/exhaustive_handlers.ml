(* Tests in which the exception and effect handlers provided by the user handle
   all possible types of [exn] and [eff] respectively. *)

let run_and_discard_effect f a =
  match f a with
  | () -> ()
  | [%effect? _, _] -> ()
;;

let run_and_discard_both f a =
  match f a with
  | () -> ()
  | [%effect? _, _] -> ()
  | exception _ -> ()
;;
