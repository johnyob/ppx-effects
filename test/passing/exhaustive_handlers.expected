open! Ppx_effects_runtime.Import
let run_and_discard_effect f a =
  Ppx_effects_runtime.Deep.match_with f a
    {
      Ppx_effects_runtime.Deep.retc = (function | () -> ());
      exnc = Ppx_effects_runtime.raise;
      effc =
        (let effc (type continue_input) =
           (let open! Ppx_effects_runtime.Deep.Open_on_rhs in
              function | _ -> Some ((fun _ -> ())) : continue_input
                                                       Ppx_effects_runtime.t
                                                       ->
                                                       ((continue_input, 
                                                          _)
                                                          Ppx_effects_runtime.Deep.continuation
                                                          -> _)
                                                         option) in
         effc)
    }
let run_and_discard_both f a =
  Ppx_effects_runtime.Deep.match_with f a
    {
      Ppx_effects_runtime.Deep.retc = (function | () -> ());
      exnc = (function | _ -> ());
      effc =
        (let effc (type continue_input) =
           (let open! Ppx_effects_runtime.Deep.Open_on_rhs in
              function | _ -> Some ((fun _ -> ())) : continue_input
                                                       Ppx_effects_runtime.t
                                                       ->
                                                       ((continue_input, 
                                                          _)
                                                          Ppx_effects_runtime.Deep.continuation
                                                          -> _)
                                                         option) in
         effc)
    }
